File: Leetcode/dfs_recursion_backtracking/22_GenerateParentheses/Solution.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res = new LinkedList<>();
        dfs(n, n, "", res);
        return res;
    }
    private void dfs(int left, int right,
                     String temp, List<String> res){
        if(left==0 && right==0){
            res.add(temp);
            return;
        }
        if(left > 0)
            dfs(left-1, right, temp+"(", res);
        if(left<right)
            dfs(left, right-1, temp+")", res);
    }
}

File: Leetcode/dfs_recursion_backtracking/22_GenerateParentheses/Solution.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        def dfs(left, right, temp):
            if left > n or right > n:
                return
            if left==n and right==n:
                res.append(temp)
                return
            if left < n:
                dfs(left+1, right, temp+"(")
            if left > right:
                dfs(left, right+1, temp+")")
        res = []
        dfs(0, 0, "")
        return res
